modules/dnn/src/op_timvx.cpp

```diff
@@ -0,0 +1,532 @@
+// This file is part of OpenCV project.
+// It is subject to the license terms in the LICENSE file found in the top-level directory
+// of this distribution and at http://opencv.org/license.html.
+//
+// Copyright (C) 2019-2021, Shenzhen Institute of Artificial Intelligence and
+// Robotics for Society, all rights reserved.
+// Third party copyrights are property of their respective owners.
+
+#include "precomp.hpp"
+#include <opencv2/dnn/shape_utils.hpp>
+#include "op_timvx.hpp"
+
+namespace cv
+{
+namespace dnn
+{
+#ifdef HAVE_TIMVX
+
+// from CPU to NPU
+bool copyToTensor(std::shared_ptr<tim::vx::Tensor> &dst, const Mat &src)
+{
+    CV_Assert(src.isContinuous());
+    if(dst->CopyDataToTensor(src.data, src.total()))
+        return true;
+    else
+        return false;
+}
+
+// from NPU to CPU
+bool copyToMat(const Mat &dst, std::shared_ptr<tim::vx::Tensor> &src)
+{
+    if(src->CopyDataFromTensor(dst.data))
+    {
+        return true;
+    }
+    else
+        return false;
+}
+
+tim::vx::ShapeType getShapeTypeFromMat(const Mat& mat, bool ifConst)
+{
+    /* Convert Mat shape to TimVx Tensor shape.
+    DataLayout in TimVx is WHCN, while NCHW in OpenCV.
+    So we do vector reverse.
+    */
+    CV_Assert(!mat.empty());
+    tim::vx::ShapeType tvInputShape;
+    auto matShape = shape(mat);
+    tvInputShape.assign(matShape.begin(), matShape.end());
+
+    if( matShape.size() > 1 )  // TODO: check when we need reverse the shape vector.
+    {
+        if(ifConst && tvInputShape.size() == 2 && tvInputShape[1] == 1)
+        {   // if bias vector, shape [n, 1] to [n].
+            tvInputShape.resize(1);
+        }else
+            std::reverse(tvInputShape.begin(), tvInputShape.end());
+    }
+    return tvInputShape;
+}
+
+bool getQuantType(const Mat& mat, int numOutput)
+{
+    CV_Assert(!mat.empty());
+    if(numOutput == -1)
+    {
+        numOutput = mat.total();
+    }
+    bool tvSymmetric = false;
+
+    for(int i =1; i < numOutput; i++)
+    {
+        if(std::abs(mat.at<float>(0) - mat.at<float>(i)) > std::numeric_limits<float>::epsilon())
+        {
+            tvSymmetric = true;
+            break;
+        }
+    }
+
+    return tvSymmetric;
+}
+
+// convert mat Depth to tensorDataType
+tim::vx::DataType dataTypeConvert(int matDepth)
+{
+    tim::vx::DataType tensorDataType;
+    switch(matDepth)
+    {
+        case 0: // CV_8U
```

Comment:
> 0

we don't need magic numbers here. Use CV_8U directly.

Reply:
Thanks, it was fixed in the latest update.